### 자료의 위치를 나타내는 대표값 : 평균, 백분위수, 사분위수(IQR), 중앙값, 절사평균, 최빈값

### 자료의 산포를 나타내는 대표값 : 사분위수 범위, 표준 편차, 분산, 절대편차

### 두 자료의 선형관계를 나타내는 값 : 공분산, 상관계수

- 자동차 연비 데이터셋에서 모델별로 연비에 대한 기술통계량

In [1]:
import pandas as pd
import numpy as np

In [13]:
path = 'C:/Users/user/Desktop/TodayLearn/TiL/Code/data/'
mycar_data = pd.read_csv('C:/Users/user/Desktop/TodayLearn/TiL/Code/data/mycars.csv')
print(mycar_data.shape)
mycar_data.head()

(234, 11)


manufacturer model  displacement  year  cylinder automatic driving  mpg  \
0         audi    a4           1.8  1999         4      auto       f   18   
1         audi    a4           1.8  1999         4    manual       f   21   
2         audi    a4           2.0  2008         4    manual       f   20   
3         audi    a4           2.0  2008         4      auto       f   21   
4         audi    a4           2.8  1999         6      auto       f   16   

   highway_mileage fuel    class  
0               29    p  compact  
1               29    p  compact  
2               31    p  compact  
3               30    p  compact  
4               26    p  compact

In [4]:
df = mycar_data[['model', 'mpg']]
df.groupby('model').describe()

mpg                                                 \
                       count       mean       std   min    25%   50%    75%   
model                                                                         
4runner 4wd              6.0  15.166667  0.752773  14.0  15.00  15.0  15.75   
a4                       7.0  18.857143  1.864454  16.0  18.00  18.0  20.50   
a4 quattro               8.0  17.125000  1.807722  15.0  15.75  17.0  18.25   
a6 quattro               3.0  16.000000  1.000000  15.0  15.50  16.0  16.50   
altima                   6.0  20.666667  1.966384  19.0  19.00  20.0  22.50   
c1500 suburban 2wd       5.0  12.800000  1.303840  11.0  12.00  13.0  14.00   
camry                    7.0  19.857143  1.463850  18.0  18.50  21.0  21.00   
camry solara             7.0  19.857143  1.772811  18.0  18.00  21.0  21.00   
caravan 2wd             11.0  15.818182  1.834022  11.0  15.50  16.0  17.00   
civic                    9.0  24.444444  1.943651  21.0  24.00  24.0  25.00   
corolla                  5.0  25.600000  1.673320  24.0  24.00  26.0  26.00   
corvette                 5.0  15.400000  0.547723  15.0  15.00  15.0  16.00   
dakota pickup 4wd        9.0  12.777778  1.986063   9.0  11.00  14.0  14.00   
durango 4wd              7.0  11.857143  1.573592   9.0  11.00  13.0  13.00   
expedition 2wd           3.0  11.333333  0.577350  11.0  11.00  11.0  11.50   
explorer 4wd             6.0  13.666667  0.816497  13.0  13.00  13.5  14.00   
f150 pickup 4wd          7.0  13.000000  1.000000  11.0  13.00  13.0  13.50   
forester awd             6.0  18.833333  0.983192  18.0  18.00  18.5  19.75   
grand cherokee 4wd       8.0  13.500000  2.507133   9.0  12.50  14.0  15.00   
grand prix               5.0  17.000000  1.000000  16.0  16.00  17.0  18.00   
gti                      5.0  20.000000  2.000000  17.0  19.00  21.0  21.00   
impreza awd              8.0  19.625000  0.744024  19.0  19.00  19.5  20.00   
jetta                    9.0  21.222222  4.867694  16.0  19.00  21.0  21.00   
k1500 tahoe 4wd          4.0  12.500000  1.732051  11.0  11.00  12.5  14.00   
land cruiser wagon 4wd   2.0  12.000000  1.414214  11.0  11.50  12.0  12.50   
malibu                   5.0  18.800000  1.923538  17.0  18.00  18.0  19.00   
maxima                   3.0  18.666667  0.577350  18.0  18.50  19.0  19.00   
mountaineer 4wd          4.0  13.250000  0.500000  13.0  13.00  13.0  13.25   
mustang                  9.0  15.888889  1.452966  14.0  15.00  15.0  17.00   
navigator 2wd            3.0  11.333333  0.577350  11.0  11.00  11.0  11.50   
new beetle               6.0  24.000000  6.511528  19.0  20.00  20.5  27.00   
passat                   7.0  18.571429  1.902379  16.0  17.50  18.0  20.00   
pathfinder 4wd           4.0  13.750000  1.258306  12.0  13.50  14.0  14.25   
ram 1500 pickup 4wd     10.0  11.400000  1.505545   9.0  11.00  11.5  12.75   
range rover              4.0  11.500000  0.577350  11.0  11.00  11.5  12.00   
sonata                   7.0  19.000000  1.414214  18.0  18.00  18.0  20.00   
tiburon                  7.0  18.285714  1.603567  16.0  17.00  19.0  19.50   
toyota tacoma 4wd        7.0  15.571429  0.786796  15.0  15.00  15.0  16.00   

                              
                         max  
model                         
4runner 4wd             16.0  
a4                      21.0  
a4 quattro              20.0  
a6 quattro              17.0  
altima                  23.0  
c1500 suburban 2wd      14.0  
camry                   21.0  
camry solara            22.0  
caravan 2wd             18.0  
civic                   28.0  
corolla                 28.0  
corvette                16.0  
dakota pickup 4wd       15.0  
durango 4wd             13.0  
expedition 2wd          12.0  
explorer 4wd            15.0  
f150 pickup 4wd         14.0  
forester awd            20.0  
grand cherokee 4wd      17.0  
grand prix              18.0  
gti                     22.0  
impreza awd             21.0  
jetta     

In [6]:
# 라이브러리
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest

In [11]:
# 차량의 도장공정에서 광택도 불량이 40% 임을 가정
# 3 대의 차량을 임의로 선태했을때, 불량대수(0,1,2,3)가 나올 확률

n = 3
for i in range(n + 1) :
    prop = stats.binom.pmf(k = i, n = n , p = 0.4).round(2)
    print("P(X = {0}) = {1:.3f}".format(i, prop))

P(X = 0) = 0.220
P(X = 1) = 0.430
P(X = 2) = 0.290
P(X = 3) = 0.060


# 귀무가설 VS 대립가설 : 추론통계..Estimation

<b>귀무가설 : </b> 설정한 가설의 진실성 확률이 적어 처음부터 버릴 것으로 예상되는 가설, 검정의 대상으로 삼는 가설

__대립가설__: 가설검정에서 진위를 검증하고자 하는 가설, 귀무가설이 기각 될때 받아들여지는 가설

 1. 강아지의 평균 수명은 13년이라고 알려져 있다. 몇 전문가는 13년이 아니라고 주장한다. 따라서 강아지의 평균수명은 실제로 어떻게 되는지를 검정하고 싶다.
 
 - 강아지의 평균 수명은 13년이다. -> 귀무가설
 - 강아지의 평균 수명은 13년이 아니다 -> 대립가설
 
 2. 전자제품의 불량률은 3%로 알려져 있다. 그런데 불량률에 대한 항의 민원 전화가 많아 불량이 3% 보다 더 클 수 있다는 소리가 나오고 있다. 이에 불량률이 실제로 어떻게 되는지 검정하고자 한다.
  - 제품 불량률은 3%보다 작거나 같다 -> 귀무가설
  - 3% 보다 크다-> 대립가설
  
 3. 어느 공장서 생산하는 제품의 분산은 7 이라고 알려져 있다. 그런데 품질관리팀의 최근 조사에 의하면, 분산은 7보다 작을 수도 있다는 의견이 나왔다. 따라서. 제품의 분산이실제로 어떠한지를 알아보기 위하여 검정을 시행하고자 한다.
 
  - 제품의 분산은 7보다 크거나 같다. -> 귀무가설
  - 제품의 분산은 7보다 작다. -> 대립가설
  
 ### 귀무가설(Null Hypothesis : 이론에 대한 가설), 대립 가설(Alternative Hypothesis : 이론에 대한 반론)
 
  __P-value__ : 귀무가설이 참이라는 가정하에 표본데이터가 귀무가설을 지지하는 확률, 이 값이 유의수준(알파)보다 작으면 귀무가설을 기각함(대립 가설을 지지).

In [15]:
df = pd.read_csv(f'{path}/data_norm.csv')
# 정규성 검정 - 정규분포에 따르는가?
# 고객만족도 평균은 76.7이 맞다(귀무가설)
x = stats.shapiro(df) # shapiro-wilk test
print('The Shapiro-wilkis Statistic is %.3f and the p-value is %.3f' % x)

The Shapiro-wilkis Statistic is 0.933 and the p-value is 0.480


In [19]:
t_result = stats.ttest_1samp(df, 76.7)
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)
print("1-Sample t-test")
print("t검정통계량 : {}".format(t))
print("p-value : {} ".format(p))

1-Sample t-test
t검정통계량 : [0.864]
p-value : [0.41] 


- alpha = 0.05, 보다 P-value(0.41, 0.480) 으로 크기 때문에 귀무가설은 적절하다. 즉, 평균은 76.7이다 -> 만족도 변화 x

In [23]:
df

sample
0    74.5
1    81.2
2    73.8
3    82.0
4    76.3
5    75.7
6    80.2
7    72.6
8    77.9
9    82.8

In [24]:
# 고객만족도 확인을 위해 A와 B고객에 대해 조사하여 다음 데이터를 얻음.
# A고객과 B고객의 모평균에 차이가 있는가?  - 귀무가설. 모평균에 차이는 없음

In [28]:
df1 = pd.read_csv(f'{path}/data_sampt1.csv')
df2 = pd.read_csv(f'{path}/data_sampt2.csv')

# 정규성 검정
x = stats.shapiro(df1)
print('Shaprirpo-Willkis검정 통계량은 %.3f, p-value는 %.3f' %x)
y = stats.shapiro(df2)
print('Shaprirpo-Willkis검정 통계량은 %.3f, p-value는 %.3f' %y)

stat, p = stats.bartlett(df1['sample'], df2['sample'])
print('등분산 검정결과 p -value 는 %.3f' %p)

# 0.05 < 0.642 , 알파보다 크므로, 정규
t_result = stats.ttest_ind(df1['sample'], df2['sample'], equal_var = True)
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)
print( "2-Sample t-test ")
print( "t검정통계량: {} " .format(t))
print( "p-value: {} " .format(p))

# p-value 가 알파 보다 작으므로, 모평균에 차이가 있다.

df1.describe()

Shaprirpo-Willkis검정 통계량은 0.909, p-value는 0.062
Shaprirpo-Willkis검정 통계량은 0.898, p-value는 0.038
등분산 검정결과 p -value 는 0.642
2-Sample t-test 
t검정통계량: -3.122 
p-value: 0.003 


sample
count  20.000000
mean    5.500000
std     1.100239
min     4.000000
25%     5.000000
50%     5.500000
75%     6.000000
max     8.000000

In [32]:
# 회사원 10명에 대해 영어점수가 학원가기 전의 점수와 학원수강 후에 영어점수의 차이가
#있는지 확인하고 싶다. 아래의 데이터를 가지고 차이가 있는지 검정하라. ( α : 5%)

# 데이터 불러오기
df = pd.read_csv(f'{path}/data_pair.csv')
# paired t-test
t_result = stats.ttest_rel(df['before'], df['after'])
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)
print( "Paired t-test ")
print( " t검정통계량 : {} " .format(t))
print( " p-value : {} " .format(p))

# 학원가기전과 학원간후 영어점수에 차이가있다. p-value < a (0.05) / 귀무가설 기각 H1 참

Paired t-test 
 t검정통계량 : -5.324 
 p-value : 0.0 


In [35]:
df.describe()

before      after
count   10.000000   10.00000
mean   690.200000  746.60000
std     75.860691   56.78067
min    589.000000  670.00000
25%    648.000000  712.00000
50%    684.000000  751.00000
75%    765.000000  790.00000
max    780.000000  810.00000

In [37]:
# 상관 분석
df = pd.read_csv(f'{path}/data_cor.csv')
df

minutes  units
0        1     23
1        2     29
2        3     49
3        4     64
4        4     74
5        5     87
6        6     96
7        6     97
8        7    109
9        8    119

In [39]:
corr, pval = stats.pearsonr(df['minutes'], df['units'])

In [41]:
# 데이터 불러오기
df = pd.read_csv(f'{path}/data_chi.csv', encoding = 'euc-kr')
# chi-square test 실행
chi, pval, dof, expected = stats.chi2_contingency(df)
print( "chi-square test ")
print( ' chisq: {0:0.3f}' .format(chi))
print( ' p: {0:0.3f}'.format(pval))
print( ' degree pf freedom: {}' .format(dof))
print( ' expected value: \n{}' .format(expected.round(3)))

chi-square test 
 chisq: 13.366
 p: 0.038
 degree pf freedom: 6
 expected value: 
[[258.333 258.333 258.333]
 [276.333 276.333 276.333]
 [230.667 230.667 230.667]
 [234.667 234.667 234.667]]
